### PR TITLE
[SPARK-14231][SQL] JSON data source infers floating-point values as a double when they do not fit in a decimal

### PR DESCRIPTION
**Pull request by HyukjinKwon**
## What changes were proposed in this pull request?

https://issues.apache.org/jira/browse/SPARK-14231

Currently, JSON data source supports to infer `DecimalType` for big numbers and `floatAsBigDecimal` option which reads floating-point values as `DecimalType`.

But there are few restrictions in Spark `DecimalType` below:
1. The precision cannot be bigger than 38.
2. scale cannot be bigger than precision. 

Currently, both restrictions are not being handled.

This PR handles the cases by inferring them as `DoubleType`. Also, the option name was changed from `floatAsBigDecimal` to `prefersDecimal` as suggested [here](https://issues.apache.org/jira/browse/SPARK-14231?focusedCommentId=15215579&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-15215579).

So, the codes below:

``` scala
def doubleRecords: RDD[String] =
  sqlContext.sparkContext.parallelize(
    s"""{"a": 1${"0" * 38}, "b": 0.01}""" ::
    s"""{"a": 2${"0" * 38}, "b": 0.02}""" :: Nil)

val jsonDF = sqlContext.read
  .option("prefersDecimal", "true")
  .json(doubleRecords)
jsonDF.printSchema()
```

produces below:
- **Before**

``` scala
org.apache.spark.sql.AnalysisException: Decimal scale (2) cannot be greater than precision (1).;
    at org.apache.spark.sql.types.DecimalType.<init>(DecimalType.scala:44)
    at org.apache.spark.sql.execution.datasources.json.InferSchema$.org$apache$spark$sql$execution$datasources$json$InferSchema$$inferField(InferSchema.scala:144)
    at org.apache.spark.sql.execution.datasources.json.InferSchema$.org$apache$spark$sql$execution$datasources$json$InferSchema$$inferField(InferSchema.scala:108)
    at 
...
```
- **After**

``` scala
root
 |-- a: double (nullable = true)
 |-- b: double (nullable = true)
```
## How was this patch tested?

Unit tests were used and `./dev/run_tests` for coding style tests.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
cc @rxin

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/InferSchema.scala [hunk 1]
```diff
@@ -135,11 +136,15 @@ private[sql] object InferSchema {
           // when we see a Java BigInteger, we use DecimalType.
           case BIG_INTEGER | BIG_DECIMAL =>
             val v = parser.getDecimalValue
-            DecimalType(v.precision(), v.scale())
+            // Creating `DecimalType` here can fail when precision is bigger than 38.
+            Try(DecimalType(v.precision(), v.scale())).getOrElse(DoubleType)
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/InferSchema.scala]: maybe use the more traditional try catch to improve performance a bit.

**Comment by rxin** [comment]
do we have a test case for data that some of them fit in a decimal and some don't?

i want to make sure we have a rule to coerce decimal + double into just double.

**Comment by HyukjinKwon** [comment]
@rxin I just replace the use of `Try` to `try/catch` and added the case you said.

**Comment by HyukjinKwon** [comment]
@rxin Could you maybe merge this if everything is okay?

**Comment by rxin** [comment]
cc @davies can you take a quick look at this to see if it is ok? thanks.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/InferSchema.scala [hunk 2]
```diff
@@ -135,11 +135,25 @@ private[sql] object InferSchema {
           // when we see a Java BigInteger, we use DecimalType.
           case BIG_INTEGER | BIG_DECIMAL =>
             val v = parser.getDecimalValue
-            DecimalType(v.precision(), v.scale())
+            try {
+              // Creating `DecimalType` here can fail when precision is bigger than 38.
+              DecimalType(v.precision(), v.scale())
```
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/InferSchema.scala]: It's reasonable to have a decimal "0.01", so we should use max of precision and scale as precision here. Also it's better  to not reply on exception for branch.

if (Math.max(v.precision(), v.scale()) <= DecimalType.MAX_MAX_PRECISION) {
  DecimalType(Math.max(v.precision(), v.scale()), v.scale())
} else {
  DoubleType
}

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala [hunk 3]
```diff
@@ -773,6 +773,45 @@ class JsonSuite extends QueryTest with SharedSQLContext with TestJsonData {
...
+      jsonDF,
+      Seq(Row(1.0E38D, 9.223372036854776E19), Row(2.0E38D, 2.0E38D))
+    )
+  }
+
+  test("Infer floating-point values as doubles when it does not fit in decimal") {
+    val jsonDF = sqlContext.read
+      .option("prefersDecimal", "true")
+      .json(doubleRecords)
+      .selectExpr("b", "d")
```
Comment (davies) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala]: It's better to check the schema of all columns

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/TestJsonData.scala [hunk 4]
```diff
@@ -214,6 +214,11 @@ private[json] trait TestJsonData {
       """{"a": {"b": 1}}""" ::
       """{"a": []}""" :: Nil)
 
+  def doubleRecords: RDD[String] =
+    sqlContext.sparkContext.parallelize(
+      s"""{"a": 1${"0" * 38}, "b": 0.01, "c": 92233720368547758070, "d": 1.01}""" ::
```
Comment (davies) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/TestJsonData.scala]: Can we have a column that have different schema (range) on different row, they got merged together finally.

**Comment by HyukjinKwon** [comment]
@davies Actually, I am not sure if I understood the last comment correctly. Would you check the tests please? I added some tests for both merged types and not-merged but just inferred types separatly.

**Comment by davies** [comment]
LGTM, merging this into master, thanks!

**Pull request closed without merging**